crates/bevy_reflect/bevy_reflect_derive/src/remote.rs

```diff
@@ -0,0 +1,474 @@
+    let wrapper_definition = generate_remote_wrapper(&ast, &remote_ty);
+
+    let mut derive_data = match ReflectDerive::from_input(
+        &ast,
+        ReflectProvenance {
+            source: ReflectImplSource::RemoteReflect,
+            trait_: ReflectTraitToImpl::Reflect,
+        },
+    ) {
+        Ok(data) => data,
+        Err(err) => return err.into_compile_error().into(),
+    };
+
+    derive_data.set_remote(Some(RemoteType::new(&remote_ty)));
+
+    let assertions = impl_assertions(&derive_data);
+    let definition_assertions = generate_remote_definition_assertions(&derive_data);
+
+    let reflect_remote_impl = impl_reflect_remote(&derive_data, &remote_ty);
+
+    let (reflect_impls, from_reflect_impl) = match derive_data {
+        ReflectDerive::Struct(struct_data) | ReflectDerive::UnitStruct(struct_data) => (
+            impls::impl_struct(&struct_data),
+            if struct_data.meta().from_reflect().should_auto_derive() {
+                Some(from_reflect::impl_struct(&struct_data))
+            } else {
+                None
+            },
+        ),
+        ReflectDerive::TupleStruct(struct_data) => (
+            impls::impl_tuple_struct(&struct_data),
+            if struct_data.meta().from_reflect().should_auto_derive() {
+                Some(from_reflect::impl_tuple_struct(&struct_data))
+            } else {
+                None
+            },
+        ),
+        ReflectDerive::Enum(enum_data) => (
+            impls::impl_enum(&enum_data),
+            if enum_data.meta().from_reflect().should_auto_derive() {
+                Some(from_reflect::impl_enum(&enum_data))
+            } else {
+                None
+            },
+        ),
+        _ => {
+            return syn::Error::new(ast.span(), "cannot reflect a remote value type")
+                .into_compile_error()
+                .into()
+        }
+    };
+
+    TokenStream::from(quote! {
+        #wrapper_definition
+
+        const _: () = {
+            #reflect_remote_impl
+
+            #reflect_impls
+
+            #from_reflect_impl
+
+            #definition_assertions
+
+            #assertions
+        };
+    })
+}
+
+/// Generates the remote wrapper type.
+///
+/// # Example
+///
+/// If the supplied remote type is `Bar<T>`, then the wrapper type— named `Foo<T>`— would look like:
+///
+/// ```
+/// # struct Bar<T>(T);
+///
+/// #[repr(transparent)]
+/// struct Foo<T>(Bar<T>);
+/// ```
+fn generate_remote_wrapper(input: &DeriveInput, remote_ty: &TypePath) -> proc_macro2::TokenStream {
+    let ident = &input.ident;
+    let vis = &input.vis;
+    let ty_generics = &input.generics;
+    let where_clause = &input.generics.where_clause;
+    let attrs = input
+        .attrs
+        .iter()
+        .filter(|attr| !attr.path().is_ident(REFLECT_ATTRIBUTE_NAME));
+
+    quote! {
+        #(#attrs)*
+        #[repr(transparent)]
+        #vis struct #ident #ty_generics (pub #remote_ty) #where_clause;
```

Comment:
We don't have to, since it is pretty much an implementation detail. But I included it so that users could have the choice to pass around the wrapper type if they so choose.

However, I can remove it for now so users don't accidentally pollute their public API. And we can add it back in if it ever comes up as a need for someone.

Reply:
there's already `as_wrapper` etc. for this. i just don't think we need to commit to this implementation as part of the public api.

Reply:
As it turns out, this is necessary. 😅

The reason is because we might define our remote type in one module, but wish to use it from another:

```rust
mod wrapper {
    #[reflect_remote(external_crate::TheirType)]
    pub struct MyType {
        pub value: String,
    }
}

mod foo {
    #[derive(Reflect)]
    struct ContainerStruct {
        #[reflect(remote = super::wrapper::MyType)]
        their_type: external_crate::TheirType,
    }
}
```

And we can't simply determine it from `TheirType` since `ReflectRemote` is only implemented for `MyType`.